polkadot/runtime/parachains/src/util.rs

```diff
@@ -98,13 +104,104 @@ pub fn take_active_subset<T: Clone>(active: &[ValidatorIndex], set: &[T]) -> Vec
 	subset
 }
 
+/// On first pass it filters out all candidates that have multiple cores assigned and no
+/// `CoreIndex` injected.
+///
+/// On second pass we filter out candidates with the same core index. This can happen if for example
+/// collators distribute collations to multiple backing groups.
+pub(crate) fn filter_elastic_scaling_candidates<T: configuration::Config + scheduler::Config>(
+	candidates: &mut Vec<BackedCandidate<T::Hash>>,
+) {
+	if !configuration::Pallet::<T>::config()
+		.node_features
+		.get(FeatureIndex::ElasticScalingCoreIndex as usize)
+		.map(|b| *b)
+		.unwrap_or(false)
+	{
+		// we don't touch the candidates, since we don't expect block producers
+		// to inject `CoreIndex`.
+		return
+	}
+
+	// A mapping from parachain to all assigned cores.
+	let mut cores_per_parachain: BTreeMap<ParaId, Vec<CoreIndex>> = BTreeMap::new();
+
+	for (core_index, para_id) in <scheduler::Pallet<T>>::scheduled_paras() {
+		cores_per_parachain.entry(para_id).or_default().push(core_index);
+	}
+
+	// We keep a candidate if the parachain has only one core assigned or if
+	// a core index is provided by block author.
+	candidates.retain(|candidate| {
+		!cores_per_parachain
+			.get(&candidate.candidate().descriptor.para_id)
+			.map(|cores| cores.len())
+			.unwrap_or(0) >
+			1 || has_core_index::<T>(candidate, true)
+	});
+
+	let mut used_cores = BTreeSet::new();
+
+	// We keep one candidate per core in case multiple candidates of same para end up backed on same
+	// core. This can be further refined to pick the candidate that has the parenthead equal
+	// to the one in storage.
+	candidates.retain(|candidate| {
+		if let Some(core_index) = candidate.assumed_core_index(true) {
+			// Drop candidate if the core was already used by a previous candidate.
+			used_cores.insert(core_index)
+		} else {
+			// This shouldn't happen, but at this point the candidate without core index is fine
+			// since we know the para:core mapping is unique.
+			true
+		}
+	})
+}
+
+// Returns `true` if the candidate contains an injected `CoreIndex`.
+fn has_core_index<T: configuration::Config + scheduler::Config>(
```

Comment:
Shouldn't this be just calling `assumed_core_index` on `BackedCandidate`?

Reply:
I think it should